File: brainfuck.py
from sys import stdin, argv

def inc(n):
	return (n+1) % 256

def dec(n):
	return 255 if n-1 < 0 else n-1

tape = []
code_pos = 0
tape_pos = 0

def run(skip=False):
	global tape_pos, code_pos, code, tape

	while tape_pos >= 0 and code_pos < len(code):
		if tape_pos >= len(tape):
			tape.append(0)

		if code[code_pos] == '[':
			code_pos += 1
			old_pos = code_pos
			while run(tape[tape_pos] == 0):
				code_pos = old_pos
		elif code[code_pos] == ']':
			return tape[tape_pos] != 0
		elif not skip:
			if code[code_pos] == '+':
				tape[tape_pos] = inc(tape[tape_pos])

			elif code[code_pos] == '-':
				tape[tape_pos] = dec(tape[tape_pos])

			elif code[code_pos] == '>':
				tape_pos += 1

			elif code[code_pos] == '<':
				tape_pos -= 1

			elif code[code_pos] == '.':
				print(chr(tape[tape_pos]), end='')

			elif code[code_pos] == ',':
				tape[tape_pos] = ord(stdin.read(1))

		code_pos += 1

	return False

if __name__ == '__main__':
	if len(argv) > 1:
		with open(argv[1]) as bfile:
			code = bfile.read()
	else:
		code = stdin.read()

	run()

File: brainfuck.nim
import os

{.push overflowchecks: off.}
proc xinc(c: var char) = inc c
proc xdec(c: var char) = dec c
{.pop.}

let code = if paramCount() > 0: readFile paramStr 1
           else: readAll stdin

var
    tape = newSeq[char]()
    codePos = 0
    tapePos = 0

proc run(skip = false): bool =
    while tapePos >= 0 and codePos < len code:
        if tapePos >= len tape:
            tape.add '\0'

        if code[codePos] == '[':
            inc codePos
            let oldPos = codePos
            while run tape[tapePos] == '\0':
                codePos = oldPos
        elif code[codePos] == ']':
            return tape[tapePos] != '\0'
        elif not skip:
            case code[codePos]
            of '+': xinc tape[tapePos]
            of '-': xdec tape[tapePos]
            of '>': inc tapePos
            of '<': dec tapePos
            of '.': write stdout, tape[tapePos]
            of ',': tape[tapePos] = readChar stdin
            else: discard

        inc codePos

discard run()